Create the app backbone with models and seeds

pczarn opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pczarn/gameroom-task/2)

<!-- Reviewable:end -->

pczarn committed:
Add associations, validations and specs

Model fields are validated for presence, uniqueness and correctness.
Factories are defined for models using faker.
Specs check associations and validations.

pczarn committed:
Bundle rubocop and postgresql

Syntax is checked with rubocop.
The database is changed to postgres in all environments.

pczarn committed:
Fix rubocop warnings

Almost all warnings are addressed.

pczarn committed:
Change all single quotes to double quotes

The guidelines say to always use double quotes.

pczarn committed:
Add and refactor specs and validations

Database indexes are added to the schema so that uniqueness will be
checked on two levels. Some code is refactored. A few specs are added.

pczarn committed:
Write database seeds

Seeds will be useful later. They don't use factories.

pczarn committed:
Add image to `Game` with carrierwave

Games are illustrated to visually tell them apart. We use carrierwave
with minimagick, because it's often used at monterail. Images and
thumbnails are resized to fit, so that they will be easier to
fit on page.
Testing leaves garbage images, so the spec_helper and uploader are
configured to delete garbage.

pczarn committed:
Initialize 4 models with factories

Models have with basic associations and fields. There is a join table
`UserTeams`, because teams will have many users and teams will have many users.
Matches have two foreign keys that point to teams, so that teams can be
reused in different matches. Rspec is configured to work with FactoryGirl.
Factories are not written yet.

pczarn commented on app/models/match.rb:
```diff
@@ -0,0 +1,42 @@
+    users_in_team_one = UserTeam.where(team: team_one).pluck(:user_id)
+    users_in_team_two = UserTeam.where(team: team_two).pluck(:user_id)
+    common_members = users_in_team_one & users_in_team_two
+    if not common_members.empty?
```
I see this should be 'unless', but I don't like unless :(

monttda replied:
unless it must be  ;). Unless is awesome like it more :p

monttda commented on app/models/match.rb:
```diff
@@ -0,0 +1,42 @@
+
+  private
+
+  # Teams not empty.
```
remove comment

monttda commented on app/models/match.rb:
```diff
@@ -0,0 +1,42 @@
+    errors.add(team_field, "Can't be empty") if UserTeam.where(team: team).empty?
+  end
+
+  # No repeated members in different teams.
```
remove comment

monttda commented on app/models/match.rb:
```diff
@@ -0,0 +1,42 @@
+# Matches are played between teams.
```
remove comment

monttda commented on app/models/match.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def no_repeated_members_across_teams
+    # There are at least two approaches. This one goes through UserTeam.
+    # Pluck team member collections to check their intersection.
```
remove comment

monttda commented on app/models/team.rb:
```diff
@@ -0,0 +1,47 @@
+# Teams represent collections of players.
+#
```
remove comment

monttda commented on app/models/team.rb:
```diff
@@ -0,0 +1,47 @@
+      errors.add(:members, "Teams with these exact members exist")
+    end
+  end
+
```
remove comments from the file

pczarn committed:
Add fixes for code review

monttda commented on app/models/team.rb:
```diff
@@ -0,0 +1,38 @@
+
+  validate :unique_member_collections_for_teams
+
+  def member_ids
```
this method should not exists  i see the method is only called in this_and_related_teams so why not just used the map there ?

pczarn replied:
its called in unique_member_collections_for_teams, too

monttda replied:
yes but `unique_member_collections_for_teams` uses map directly not members_id so remove this method

pczarn replied:
then how can I use map to keep the code clear? I'll write `if contains_duplicates?(this_and_related_teams.map { |team| team.members.map(&:id) })`

pczarn replied:
I'm going to add a method `this_and_related_team_member_ids`

monttda commented on app/models/team.rb:
```diff
@@ -0,0 +1,38 @@
+
+private
+
+def contain_duplicates?(ary)
```
what is this code doing and why is it out of the class ?

monttda commented on app/uploaders/image_uploader.rb:
```diff
@@ -0,0 +1,53 @@
+  # This is a sensible default for uploaders that are meant to be mounted:
+  def store_dir
+    base = if Rails.env.test?
+      "#{Rails.root}/spec/support/uploads"
```
maybe a one line if  condition ?  (if true ): (if false)

monttda commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+  describe "associations" do
+    it "includes matches" do
+      expect(game.match).not_to be_empty
```
I think this test does not represent behaviour of the Game as a unit thus it shouldn't be here

monttda commented on spec/models/game_spec.rb:
```diff
@@ -0,0 +1,40 @@
+      before { create(:game, name: "our_game") }
+
+      let(:built_game) { build(:game, name: "") }
+      it "is present" do
```
I do not understand what are you trying to test here

pczarn replied:
the test name is bad. it should be something like "cannot be blank"
and the `before` block above doesn't matter here

monttda replied:
the test name then should be `when attribute X is missing` we describe the condition on the test the result ( not being valid ) is described by the expect inside the it

monttda commented on spec/models/match_spec.rb:
```diff
@@ -0,0 +1,77 @@
+  describe "score" do
+    let(:match) { build(:match, team_one_score: 3) }
+    it "is present" do
+      expect(match.team_one_score).to eq(3)
```
this test does not give us value. there should be a test of when the model should be valid ( not verify that the presence of an attribute when it's valid)

pczarn replied:
so is this test ok, or should it be removed? I'll add another test for validity

monttda commented on spec/models/match_spec.rb:
```diff
@@ -0,0 +1,77 @@
+
+  context "with valid data" do
+    context ".no_repeated_members_across_teams" do
+      let(:match) { build(:match, team_one: team_a, team_two: team_b) }
```
you have a describe validations below but this validation is here ? it should be inside the validations scope. Also group all the possible results of one method inside it's own scope  not across the file

pczarn replied:
corrected

monttda commented on spec/models/match_spec.rb:
```diff
@@ -0,0 +1,77 @@
+  describe "validations" do
+    context ".played_at" do
+      let(:match) { build(:match, played_at: nil) }
+      it "is not missing" do
```
this message doesn't give me a feedback as a dev. it says it is not missing but has played_at nil so what is not missing?

monttda commented on spec/models/match_spec.rb:
```diff
@@ -0,0 +1,77 @@
+    context ".no_repeated_members_across_teams" do
+      let(:match) { build(:match, team_one: team_a, team_two: team_b) }
+      it "can have teams with distinct players" do
+        expect(match).to be_valid
```
where is the test when this is supposed to be invalid ?

monttda commented on app/models/user_team.rb:
```diff
@@ -0,0 +1,9 @@
+# Join table between users and teams.
+#
```
remove comments

pczarn commented:
Review status: 0 of 35 files reviewed at latest revision, 16 unresolved discussions.

---

_[app/models/team.rb, line 36 at r2](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbV-r2-36:-KRTONnys5IwkiOcjkjV:bw0nslo) ([raw file](https://github.com/pczarn/gameroom-task/blob/a9c600dfce07093a81586d541e9aeb7db9d6a5f7/app/models/team.rb#L36)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> what is this code doing and why is it out of the class ?
> </details>
> do you mean it should be in the class? should it be commented?

---

_[app/uploaders/image_uploader.rb, line 16 at r2](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbY-r2-16:-KRTNr13sO8dM-iBLdCL:bb6un1j) ([raw file](https://github.com/pczarn/gameroom-task/blob/a9c600dfce07093a81586d541e9aeb7db9d6a5f7/app/uploaders/image_uploader.rb#L16)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> maybe a one line if  condition ?  (if true ): (if false)
> </details>
> Done

---

_[spec/models/game_spec.rb, line 20 at r1](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbk-r1-20:-KRTNtcYNN_8WfcgwHRl:bb6un1j) ([raw file](https://github.com/pczarn/gameroom-task/blob/b2ca9d021bcbd0142bb465a8a99245568eb18a68/spec/models/game_spec.rb#L20)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> I think this test does not represent behaviour of the Game as a unit thus it shouldn't be here
> </details>
> Done

---

_[spec/models/game_spec.rb, line 30 at r1](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbk-r1-30:-KRTOBM3LwC2qAZ3oux-:bb6un1j) ([raw file](https://github.com/pczarn/gameroom-task/blob/b2ca9d021bcbd0142bb465a8a99245568eb18a68/spec/models/game_spec.rb#L30)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> the test name then should be `when attribute X is missing` we describe the condition on the test the result ( not being valid ) is described by the expect inside the it
> </details>
> Done

---

_[spec/models/match_spec.rb, line 41 at r1](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbl-r1-41:-KRTOE1m9LuD6pxx9JKL:b-yigp24) ([raw file](https://github.com/pczarn/gameroom-task/blob/b2ca9d021bcbd0142bb465a8a99245568eb18a68/spec/models/match_spec.rb#L41)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> this message doesn't give me a feedback as a dev. it says it is not missing but has played_at nil so what is not missing?
> </details>
> corrected

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 22 of 35 files at r1, 1 of 4 files at r3, 2 of 2 files at r4, 4 of 5 files at r7.
Review status: 29 of 35 files reviewed at latest revision, 29 unresolved discussions.

---

_[app/models/game.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTOs4rSyhKAXmMtCBY:-KRTOs4rSyhKAXmMtCBZ:b-2znfk8) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/game.rb#L4)):_

> ``` Ruby
> #
> class Game < ApplicationRecord
>   has_many :match
> ```

should be `matches`

---

_[app/models/match.rb, line 36 at r1](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTMOdh2-sjBQLOqPbU-r1-36:-KRTUE-glVnWsjDHrWGL:bc8fq4f) ([raw file](https://github.com/pczarn/gameroom-task/blob/b2ca9d021bcbd0142bb465a8a99245568eb18a68/app/models/match.rb#L36)):_
<details><summary>

<i>Previously, monttda (Daniel Montenegro) wrote…</i></summary>



> unless it must be  ;). Unless is awesome like it more :p
> </details>
> I agree with Daniel, but if you wish you can also write `if common_members.any?`

---

_[app/models/match.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTRaEi4Z97DmCrAH7v:-KRTRaEi4Z97DmCrAH7w:blef0w3) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L4)):_

> ``` Ruby
>   belongs_to :game
>   belongs_to :team_one, class_name: Team
>   belongs_to :team_two, class_name: Team
> ```

Presence of `team_one` and `team_two` should be also validated.

---

_[app/models/match.rb, line 19 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTSbFF4j3_ZMWV4CtP:-KRTSbFF4j3_ZMWV4CtQ:b-i10bn2) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L19)):_

> ``` Ruby
>   end
> 
>   def team_not_empty(team_field)
> ```

This validation should go to `Team` model.

---

_[app/models/match.rb, line 25 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTTocEC7SItp17F73Y:-KRTTocEC7SItp17F73Z:b-4gtvho) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L25)):_

> ``` Ruby
> 
>   def no_repeated_members_across_teams
>     users_in_team_one = UserTeam.where(team: team_one).pluck(:user_id)
> ```

you can use `team_one.member_ids`

---

_[app/models/user.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTVmENBz3OTUOeBVqo:-KRTVmENBz3OTUOeBVqp:b-x5pszd) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/user.rb#L5)):_

> ``` Ruby
>   has_many :teams, through: :user_teams
> 
>   validates_associated :user_teams
> ```

why do you need that?

---

_[db/migrate/20160905113638_create_games.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTWm5Q2Vq9NL-0Qw3l:-KRTWm5RRVPsyVA6HoQM:b-c1pruz) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113638_create_games.rb#L4)):_

> ``` Ruby
>   def change
>     create_table :games do |t|
>       t.string :name
> ```

`name` should be required

---

_[db/migrate/20160905113705_create_users.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTWrzLIEKWLebIF2-T:-KRTWrzLIEKWLebIF2-U:b-c1pruz) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113705_create_users.rb#L4)):_

> ``` Ruby
>   def change
>     create_table :users do |t|
>       t.string :name
> ```

`name` should be required

---

_[db/migrate/20160905113715_create_teams.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTX1MCIuY9w44YHNcH:-KRTX1MCIuY9w44YHNcI:bwfwn3y) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113715_create_teams.rb#L4)):_

> ``` Ruby
>   def change
>     create_table :teams do |t|
>       t.string :name
> ```

`name` should be required here also I guess, unless you have a counterargument :)

---

_[db/migrate/20160905113939_create_matches.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTXEcVddgKsQ1V8F0w:-KRTXEcVddgKsQ1V8F0x:bchmuno) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113939_create_matches.rb#L5)):_

> ``` Ruby
>     create_table :matches do |t|
>       t.datetime :played_at, null: false
>       t.references :game, foreign_key: true
> ```

should be required

---

_[db/migrate/20160905113939_create_matches.rb, line 7 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTXHrwljGF7MteVXFg:-KRTXHrwljGF7MteVXFh:b-bxzdaz) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113939_create_matches.rb#L7)):_

> ``` Ruby
>       t.references :game, foreign_key: true
>       t.references :team_one, references: :team
>       t.references :team_two, references: :team
> ```

teams should be required and why didn't you add `foreign_key` here also, like you did for `game`?

---

_[db/migrate/20160905113939_create_matches.rb, line 9 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTXO7pmaCO37X_O6ni:-KRTXO7pmaCO37X_O6nj:b-p72sm8) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113939_create_matches.rb#L9)):_

> ``` Ruby
>       t.references :team_two, references: :team
>       t.integer :team_one_score, null: true
>       t.integer :team_two_score, null: true
> ```

you don't need to pass `null: true`, it's the default

---

_[db/migrate/20160905113958_create_user_teams.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTY7EluUpff1UVpFb3:-KRTY7Ems0Uy91eXKRS1:b-fe22th) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113958_create_user_teams.rb#L5)):_

> ``` Ruby
>     create_table :user_teams do |t|
>       t.references :user, foreign_key: true
>       t.references :team, foreign_key: true
> ```

those two should be required

---

_[spec/factories/teams.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTYuPgw5bF37Ze9Xqs:-KRTYuPgw5bF37Ze9Xqt:bxdfmqe) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/factories/teams.rb#L5)):_

> ``` Ruby
>     sequence(:name) { |n| "Team #{Faker::Team.creature} #{n}" }
> 
>     # can there be 'empty_team' instead?
> ```

the answer is - no, team must have at least one user, always.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 2 of 35 files at r1, 1 of 4 files at r3, 1 of 5 files at r7.
Review status: 33 of 35 files reviewed at latest revision, 33 unresolved discussions.

---

_[spec/models/game_spec.rb, line 6 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRT_0hyjj6kNHn2IwPm:-KRT_0hzEsP5S1w5Su3P:bclysa8) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/game_spec.rb#L6)):_

> ``` Ruby
>   let(:game) { create(:game_with_matches) }
> 
>   describe ".name" do
> ```

`name` is an instance method, so accordingly to the convention it should be described as `#name` (`.name` would correspond to a class method), see http://betterspecs.org/#describe

---

_[spec/models/game_spec.rb, line 14 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTa048YgpI3KPZGA5B:-KRTa048YgpI3KPZGA5C:bz24ytw) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/game_spec.rb#L14)):_

> ``` Ruby
>   describe ".image" do
>     it "is present" do
>       expect(game.image).to be_present
> ```

for that kind of tests I recommend using this gem: (but with the `expect` syntax) https://github.com/thoughtbot/shoulda-matchers#should-vs-is_expectedto

---

_[spec/models/match_spec.rb, line 12 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTasZKDjDFJy9WIFMO:-KRTasZLashrjRz3BelA:bpm9m8) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/match_spec.rb#L12)):_

> ``` Ruby
>       let(:match) { build(:ongoing_match) }
>       it "is absent" do
>         expect(match.team_one_score).to be_nil
> ```

this test doesn't make much sense because you are testing that you in fact haven't provided the score. It would have sense if there would be some kind of a logic that would modify the score somehow, but there isn't. In this case it's more like testing if the `ongoing_match` factory was properly initialized.

---

_[spec/models/user_team_spec.rb, line 3 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTcmsoeoaVPMR8dyE2:-KRTcmspoMUhnkIbg4Mg:b-y7ax7t) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/user_team_spec.rb#L3)):_

> ``` Ruby
> require "rails_helper"
> 
> # should this go to team_spec?
> ```

No, it should be here because the uniqueness of `user` and `team` is defined in `UserTeam` model.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

pczarn commented:
Review status: 33 of 35 files reviewed at latest revision, 33 unresolved discussions.

---

_[app/models/game.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTOs4rSyhKAXmMtCBY:-KRTmaaEqI-JFW8_MJm3:bb7f2mf) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/game.rb#L4)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> should be `matches`
> </details>
> done

---

_[app/models/match.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTRaEi4Z97DmCrAH7v:-KRTm_6sOWc5Ja_7c8UM:bb7f2mf) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L4)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> Presence of `team_one` and `team_two` should be also validated.
> </details>
> done

---

_[app/models/match.rb, line 19 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTSbFF4j3_ZMWV4CtP:-KRTmdetc9UN4UKEPSIY:bu8p9e5) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L19)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> This validation should go to `Team` model.
> </details>
> moving this validation to `Team` gives me "Team can't be empty" in most tests

---

_[app/models/match.rb, line 25 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTTocEC7SItp17F73Y:-KRTljy9cS3vHbD-Fu9g:bh8m6yz) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/match.rb#L25)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> you can use `team_one.member_ids`
> </details>
> actually, `member_ids` break my use of a factory which creates `UserTeam`, but doesn't set `members`.

---

_[app/models/user.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTVmENBz3OTUOeBVqo:-KRTq2Pl-wsTqY3SmOmN:b-5xixbi) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/app/models/user.rb#L5)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> why do you need that?
> </details>
> looks like I don't need that

---

_[db/migrate/20160905113715_create_teams.rb, line 4 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTX1MCIuY9w44YHNcH:-KRTmzyPBZVhPYSahwvK:bw3dlel) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113715_create_teams.rb#L4)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> `name` should be required here also I guess, unless you have a counterargument :)
> </details>
> I have a counterargument. Teams can be identified by their members, so a team with one person can be named with that persons name. A team with Seba and Ryszard can be called "Seba and Ryszard" by default

---

_[db/migrate/20160905113939_create_matches.rb, line 7 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTXHrwljGF7MteVXFg:-KRTnXdrvjIa1DspoTPs:byudv1p) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113939_create_matches.rb#L7)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> teams should be required and why didn't you add `foreign_key` here also, like you did for `game`?
> </details>
> I had to remove `foreign_key`, it didn't work on postgresql

---

_[db/migrate/20160905113939_create_matches.rb, line 9 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTXO7pmaCO37X_O6ni:-KRTneKuqGwJSXanChYP:b2ni9bh) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/db/migrate/20160905113939_create_matches.rb#L9)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> you don't need to pass `null: true`, it's the default
> </details>
> ok.

---

_[spec/factories/teams.rb, line 5 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTYuPgw5bF37Ze9Xqs:-KRTw6HBfshhJmHgV-mN:b2ni9bh) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/factories/teams.rb#L5)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> the answer is - no, team must have at least one user, always.
> </details>
> ok.

---

_[spec/models/game_spec.rb, line 6 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRT_0hyjj6kNHn2IwPm:-KRTw8ZhB99zh7lxvd2n:b9viezi) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/game_spec.rb#L6)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> `name` is an instance method, so accordingly to the convention it should be described as `#name` (`.name` would correspond to a class method), see http://betterspecs.org/#describe
> </details>
> ok. so everywhere I have a dot, should I change it to `#`?

---

_[spec/models/game_spec.rb, line 14 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTa048YgpI3KPZGA5B:-KRTyNzW0C8PbisYkxCK:bcwowkt) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/game_spec.rb#L14)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> for that kind of tests I recommend using this gem: (but with the `expect` syntax) https://github.com/thoughtbot/shoulda-matchers#should-vs-is_expectedto
> </details>
> I'll try it

---

_[spec/models/match_spec.rb, line 12 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTasZKDjDFJy9WIFMO:-KRTwhYcg5TSuiacV0WT:b-ynwl21) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/match_spec.rb#L12)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> this test doesn't make much sense because you are testing that you in fact haven't provided the score. It would have sense if there would be some kind of a logic that would modify the score somehow, but there isn't. In this case it's more like testing if the `ongoing_match` factory was properly initialized.
> </details>
> I removed it.

---

_[spec/models/user_team_spec.rb, line 3 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRTcmsoeoaVPMR8dyE2:-KRTwnJOt3T8J3KhwR9-:b-t8vyvx) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/user_team_spec.rb#L3)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> No, it should be here because the uniqueness of `user` and `team` is defined in `UserTeam` model.
> </details>
> I think this comment was for something else that I already moved to team_spec

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 10 of 14 files at r8.
Review status: 31 of 35 files reviewed at latest revision, 24 unresolved discussions.

---

_[spec/models/game_spec.rb, line 6 at r7](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRT_0hyjj6kNHn2IwPm:-KRXhn-LSzIhFhiwuYqb:b-hmils0) ([raw file](https://github.com/pczarn/gameroom-task/blob/b1ade92a40bca8d8b1fa1e0318933a8a17c67cf3/spec/models/game_spec.rb#L6)):_
<details><summary>

<i>Previously, pczarn (Piotr Czarnecki) wrote…</i></summary>



> ok. so everywhere I have a dot, should I change it to `#`?
> </details>
> yup :)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

pczarn commented:
Review status: 28 of 35 files reviewed at latest revision, 25 unresolved discussions.

---

_[spec/factories/teams.rb, line 10 at r9](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRYFRCrhVG9SHghVaUm:-KRYFRCrhVG9SHghVaUn:b9xcm0z) ([raw file](https://github.com/pczarn/gameroom-task/blob/8f25cee5eb16ad667951cb942a41530e9a69b85e/spec/factories/teams.rb#L10)):_

> ``` Ruby
> 
>     after(:build) do |team, evaluator|
>       if team.members.empty?
> ```

here, I'm trying to let tests set members. Without this, `team.members` are overwritten on the next line. I can remove this and set members in before in tests. Or I can add an `empty_team` factory.

---

_[spec/models/team_spec.rb, line 20 at r9](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRYFbnnUBmKvzE5Wche:-KRYFbnnUBmKvzE5Wchf:bxzwz2) ([raw file](https://github.com/pczarn/gameroom-task/blob/8f25cee5eb16ad667951cb942a41530e9a69b85e/spec/models/team_spec.rb#L20)):_

> ``` Ruby
>       context "when members are empty" do
>         let(:team) { build(:team) }
>         before { team.members = [] }
> ```

this looks bad

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 1 of 14 files at r8, 5 of 6 files at r9.
Review status: 34 of 35 files reviewed at latest revision, 22 unresolved discussions.

---

_[spec/factories/teams.rb, line 10 at r9](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRYFRCrhVG9SHghVaUm:-KRYSJUkvzAtTYV0NGuP:b-y6jnfp) ([raw file](https://github.com/pczarn/gameroom-task/blob/8f25cee5eb16ad667951cb942a41530e9a69b85e/spec/factories/teams.rb#L10)):_
<details><summary>

<i>Previously, pczarn (Piotr Czarnecki) wrote…</i></summary>



> here, I'm trying to let tests set members. Without this, `team.members` are overwritten on the next line. I can remove this and set members in before in tests. Or I can add an `empty_team` factory.
> </details>
> I would leave it as it is. But you can simply write:
> 
> ``` ruby
> team.members << create(:user)
> ```

---

_[spec/models/team_spec.rb, line 20 at r9](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRYFbnnUBmKvzE5Wche:-KRYTgX2lvnKu_yfGx34:be6agz3) ([raw file](https://github.com/pczarn/gameroom-task/blob/8f25cee5eb16ad667951cb942a41530e9a69b85e/spec/models/team_spec.rb#L20)):_
<details><summary>

<i>Previously, pczarn (Piotr Czarnecki) wrote…</i></summary>



> this looks bad
> </details>
> why? It's ok, because the proper behaviour should be with members assigned. Here we wants to check if that is validated, so we want to modify the proper instance in a way that will raise an error. In before block you are doing exactly what the test description says - and that is the right way of testing :)

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 1 of 6 files at r9.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

pczarn commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

_[spec/factories/teams.rb, line 10 at r9](https://reviewable.io:443/reviews/pczarn/gameroom-task/2#-KRYFRCrhVG9SHghVaUm:-KRYaEU5XE39yB_9f3Wi:bb7f2mf) ([raw file](https://github.com/pczarn/gameroom-task/blob/8f25cee5eb16ad667951cb942a41530e9a69b85e/spec/factories/teams.rb#L10)):_
<details><summary>

<i>Previously, jotka87 (Kamila Straus) wrote…</i></summary>



> I would leave it as it is. But you can simply write:
> 
> ``` ruby
> team.members << create(:user)
> ```
> 
> </details>
> done

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->

jotka87 commented:
Reviewed 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pczarn/gameroom-task/2)_

<!-- Sent from Reviewable.io -->
